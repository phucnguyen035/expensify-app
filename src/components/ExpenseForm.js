import React, { Component } from 'react';
import PropTypes from 'prop-types';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

export default class ExpenseForm extends Component {
  constructor(props) {
    super(props);

    this.state = {
      desc: props.desc ? props.desc : '',
      note: props.note ? props.note : '',
      amount: props.amount ? (props.amount / 100).toString() : '',
      createdAt: props.createdAt ? moment(props.createdAt) : moment(),
      focused: false,
      error: ''
    };
  }

  onDescChange = (e) => {
    const desc = e.target.value;

    this.setState(() => ({ desc }));
  };

  onNoteChange = (e) => {
    const note = e.target.value;

    this.setState(() => ({ note }));
  };

  onAmountChange = (e) => {
    const amount = e.target.value;
    // Match amount with number only with 2 decimal places
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  };

  onDateChange = (createdAt) => {
    if (createdAt) {
      this.setState(() => ({ createdAt }));
    }
  };

  onFocusChange = ({ focused }) => {
    this.setState(() => ({ focused }));
  };

  onFormSubmit = (e) => {
    e.preventDefault();
    if (!this.state.desc || !this.state.amount) {
      this.setState(() => ({
        error: 'Please provide description and amount.'
      }));
    } else {
      this.setState(() => ({ error: '' }));
      this.props.onSubmit({
        desc: this.state.desc,
        amount: parseFloat(this.state.amount, 10) * 100,
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note
      });
    }
  };

  render() {
    return (
      <form className="form" onSubmit={this.onFormSubmit}>
        {this.state.error && <p className="form__error">{this.state.error}</p>}

        <input
          type="text"
          className="text-input"
          value={this.state.desc}
          placeholder="Description"
          onChange={this.onDescChange}
        />

        <input
          type="text"
          className="text-input"
          value={this.state.amount}
          placeholder="Amount"
          onChange={this.onAmountChange}
        />

        <SingleDatePicker
          date={this.state.createdAt}
          onDateChange={this.onDateChange}
          focused={this.state.focused}
          onFocusChange={this.onFocusChange}
          numberOfMonths={1}
          isOutsideRange={() => false}
          displayFormat="D/M/YYYY"
        />

        <textarea
          className="text-area"
          value={this.state.note}
          onChange={this.onNoteChange}
          placeholder="Add a note for this expense (optional)"
        />

        <div>
          <button className="button">Save</button>
        </div>
      </form>
    );
  }
}

ExpenseForm.propTypes = {
  desc: PropTypes.string,
  amount: PropTypes.number,
  note: PropTypes.string,
  onSubmit: PropTypes.func,
  createdAt: PropTypes.number
};

ExpenseForm.defaultProps = undefined;
