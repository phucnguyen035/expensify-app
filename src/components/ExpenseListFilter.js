import React, { Fragment, Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import {
  setTextFilters,
  sortByAmount,
  sortByDate,
  setStartDate,
  setEndDate
} from '../actions/filters';

const mapStateToProps = state => ({
  filters: state.filters
});
const mapDispatchtoProps = dispatch => ({
  setTextFilters: text => dispatch(setTextFilters(text)),
  sortByAmount: () => dispatch(sortByAmount()),
  sortByDate: () => dispatch(sortByDate()),
  setStartDate: startDate => dispatch(setStartDate(startDate)),
  setEndDate: endDate => dispatch(setEndDate(endDate))
});

export class ExpenseListFilter extends Component {
  state = {
    focusedInput: null
  };

  // Sort function
  onSortChange = (sortValue) => {
    if (sortValue === 'date') {
      this.props.sortByDate();
    } else if (sortValue === 'amount') {
      this.props.sortByAmount();
    }
  };

  onTextChange = (text) => {
    this.props.setTextFilters(text);
  };
  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };

  focusedInput = (focusedInput) => {
    this.setState(() => ({ focusedInput }));
  };

  render() {
    return (
      <Fragment>
        <input
          type="text"
          value={this.props.filters.text}
          onChange={e => this.onTextChange(e.target.value)}
        />
        <select value={this.props.filters.sortBy} onChange={e => this.onSortChange(e.target.value)}>
          <option value="date">Date</option>
          <option value="amount">Amount</option>
        </select>

        <DateRangePicker
          startDate={this.props.filters.startDate}
          // startDateId="startDate_filters"
          endDate={this.props.filters.endDate}
          // endDateId="endDate_filters"
          onDatesChange={this.onDatesChange}
          focusedInput={this.state.focusedInput}
          onFocusChange={this.focusedInput}
          isOutsideRange={() => false}
          numberOfMonths={1}
          showClearDates
          displayFormat="DD/MM/YYYY"
        />
      </Fragment>
    );
  }
}

ExpenseListFilter.propTypes = {
  filters: PropTypes.objectOf(PropTypes.any),
  sortByDate: PropTypes.func,
  sortByAmount: PropTypes.func,
  setTextFilters: PropTypes.func,
  setStartDate: PropTypes.func,
  setEndDate: PropTypes.func
};

ExpenseListFilter.defaultProps = undefined;

export default connect(
  mapStateToProps,
  mapDispatchtoProps
)(ExpenseListFilter);
