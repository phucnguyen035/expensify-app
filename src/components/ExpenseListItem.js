import React from 'react';
import PropTypes from 'prop-types';
import moment from 'moment';
import numeral from 'numeral';
import { Link } from 'react-router-dom';

export const ExpenseListItem = ({ amount, createdAt, desc, id }) => (
  <Link to={`/edit/${id}`} className="list-item">
    <div>
      <h3 className="list-item__title">{desc}</h3>
      <span className="list-item__subtitle">{moment(createdAt).format('DD/MM/YYYY')}</span>
    </div>

    <h3 className="list-item__data">{numeral(amount / 100).format('$0,0.00')}</h3>
  </Link>
);

ExpenseListItem.propTypes = {
  amount: PropTypes.number.isRequired,
  createdAt: PropTypes.number.isRequired,
  desc: PropTypes.string.isRequired,
  id: PropTypes.string.isRequired
};

export default ExpenseListItem;
