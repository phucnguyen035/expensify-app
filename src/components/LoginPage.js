import React from 'react';

const LoginPage = () => (
  <div>
    <h3>Login</h3>
  </div>
);

export default LoginPage;
