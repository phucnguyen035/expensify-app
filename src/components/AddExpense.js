import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startAddExpense } from '../actions/expenses';

const mapDispatchToProps = dispatch => ({
  startAddExpense: expense => dispatch(startAddExpense(expense))
});

export class AddExpense extends Component {
  onSubmit = (expense) => {
    this.props.startAddExpense(expense);
    this.props.history.push('/dashboard');
  };

  render() {
    return (
      <div>
        <div className="page-header">
          <div className="container">
            <h2 className="page-header__title">Add expense</h2>
          </div>
        </div>

        <div className="container">
          <button className="button button--back" onClick={this.props.history.goBack}>
            Back
          </button>
          <ExpenseForm onSubmit={this.onSubmit} />
        </div>
      </div>
    );
  }
}

AddExpense.propTypes = {
  startAddExpense: PropTypes.func,
  history: PropTypes.shape({
    goBack: PropTypes.func,
    push: PropTypes.func
  }).isRequired
};

AddExpense.defaultProps = undefined;

export default connect(
  undefined,
  mapDispatchToProps
)(AddExpense);
