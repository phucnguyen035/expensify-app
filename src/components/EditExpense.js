import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';

const mapStateToProps = (state, props) => ({
  expense: state.expenses.find(expense => expense.id === props.match.params.id)
});
const mapDispatchToProps = dispatch => ({
  startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
  startRemoveExpense: id => dispatch(startRemoveExpense(id))
});

export class EditExpense extends Component {
  onSubmit = (expense) => {
    this.props.startEditExpense(this.props.expense.id, expense);
    this.props.history.push('/dashboard');
  };
  onClick = () => {
    this.props.startRemoveExpense({ id: this.props.expense.id });
    this.props.history.push('/dashboard');
  };

  render() {
    return (
      <div>
        <ExpenseForm {...this.props.expense} onSubmit={this.onSubmit} />
        <button onClick={this.onClick}>Remove</button>
      </div>
    );
  }
}

EditExpense.propTypes = {
  startEditExpense: PropTypes.func.isRequired,
  startRemoveExpense: PropTypes.func.isRequired,

  history: PropTypes.shape({
    push: PropTypes.func
  }).isRequired,

  expense: PropTypes.objectOf(PropTypes.any).isRequired
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(EditExpense);
