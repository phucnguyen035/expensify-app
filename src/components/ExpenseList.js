import React, { Fragment } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import getVisibleExpenses from '../selectors/expenses';
import { ExpenseListItem } from './ExpenseListItem';
import ExpenseListFilter from './ExpenseListFilter';

const mapStateToProps = state => ({
  expenses: getVisibleExpenses(state.expenses, state.filters)
});

export const ExpenseList = props => (
  <Fragment>
    <ExpenseListFilter />

    {props.expenses.length === 0 ? (
      <p>No expenses</p>
    ) : (
      props.expenses.map(expense => <ExpenseListItem {...expense} key={expense.id} />)
    )}
  </Fragment>
);

ExpenseList.propTypes = {
  expenses: PropTypes.arrayOf(PropTypes.any)
};

ExpenseList.defaultProps = undefined;

export default connect(mapStateToProps)(ExpenseList);
