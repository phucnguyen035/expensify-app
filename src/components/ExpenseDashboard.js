import React from 'react';
import ExpenseList from './ExpenseList';
import ExpensesSummary from './ExpensesSummary';
import ExpenseListFilter from './ExpenseListFilter';

const ExpenseDashboard = () => (
  <div>
    <ExpensesSummary />
    <ExpenseListFilter />
    <ExpenseList />
  </div>
);

export default ExpenseDashboard;
