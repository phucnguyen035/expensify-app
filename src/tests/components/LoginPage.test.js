import React from 'react';
import { shallow } from 'enzyme';
import { LoginPage } from '../../components/LoginPage';

describe('Login page', () => {
  test('should render correctly', () => {
    const wrapper = shallow(<LoginPage />);

    expect(wrapper).toMatchSnapshot();
  });

  test('should call loginHandler on button click', () => {
    const loginSpy = jest.fn();
    const wrapper = shallow(<LoginPage loginHandler={loginSpy} />);

    wrapper.find('button').simulate('click');
    expect(loginSpy).toHaveBeenCalled();
  });
});
