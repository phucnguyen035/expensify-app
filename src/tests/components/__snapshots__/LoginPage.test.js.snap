// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login page should render correctly 1`] = `
<div>
  <h3>
    Login
  </h3>
  <button>
    Login
  </button>
</div>
`;
